Applications/Spire/Source/Ui/InfoTip.cpp

```diff
@@ -0,0 +1,277 @@
+}
+
+void InfoTip::leaveEvent(QEvent* event) {
+  fade_out();
+  QWidget::leaveEvent(event);
+}
+
+void InfoTip::paintEvent(QPaintEvent* event) {
+  auto painter = QPainter(this);
+  painter.drawPixmap(0, 0, render_background());
+}
+
+void InfoTip::set_interactive(bool is_interactive) {
+  m_is_interactive = is_interactive;
+}
+
+void InfoTip::fade_in() {
+  fade_window(this, false, FADE_IN_MS);
+}
+
+void InfoTip::fade_out() {
+  auto animation = fade_window(this, true, FADE_IN_MS);
+  connect(animation, &QPropertyAnimation::finished, [&] {
+    if(!parentWidget()->underMouse()) {
+      m_show_timer.stop();
+      hide();
+    }
+  });
+}
+
+QPainterPath InfoTip::get_arrow_path() const {
+  auto path = QPainterPath();
+  auto polygon = [&] () -> QPolygonF {
+    auto margins = get_margins();
+    auto left_x = ARROW_X_POSITION() + DROP_SHADOW_WIDTH();
+    auto tip_x = left_x + (ARROW_SIZE().width() / 2);
+    auto right_x = left_x + ARROW_SIZE().width();
+    auto orientation = get_orientation();
+    if(orientation == Orientation::TOP_LEFT ||
+        orientation == Orientation::TOP_RIGHT) {
+      return QVector<QPoint>({{left_x, height() - margins.bottom()},
+        {tip_x, height() - Y_OFFSET()},
+        {right_x, height() - margins.bottom()}});
+    }
+    return QVector<QPoint>({{left_x, margins.top()}, {tip_x, Y_OFFSET()},
+      {right_x, margins.top()}});
+  }();
+  if(get_body_orientation() == BodyOrientation::LEFT) {
+    polygon.translate(width() - (2 * (ARROW_X_POSITION() +
+      DROP_SHADOW_WIDTH())) - ARROW_SIZE().width(), 0);
+  }
+  path.addPolygon(polygon);
+  return path;
+}
+
+InfoTip::BodyOrientation InfoTip::get_body_orientation() const {
+  auto parent_position = parentWidget()->mapToGlobal(
+    parentWidget()->rect().bottomLeft());
+  auto screen_geometry =
+    get_current_screen(parent_position)->availableGeometry();
+  if(parent_position.x() + width() >
+      screen_geometry.x() + screen_geometry.width()) {
+    return BodyOrientation::LEFT;
+  }
+  return BodyOrientation::RIGHT;
+}
+
+QScreen* InfoTip::get_current_screen(const QPoint& point) const {
+  if(auto screen = QGuiApplication::screenAt(point)) {
+    return screen;
+  }
+  return parentWidget()->screen();
+}
+
+QMargins InfoTip::get_margins() const {
+  auto orientation = get_orientation();
+  if(orientation == Orientation::TOP_LEFT ||
+      orientation == Orientation::TOP_RIGHT) {
+    return {DROP_SHADOW_WIDTH(), DROP_SHADOW_HEIGHT(), DROP_SHADOW_WIDTH(),
+      Y_OFFSET() + ARROW_SIZE().height()};
+  }
+  return {DROP_SHADOW_WIDTH(), Y_OFFSET() + ARROW_SIZE().height(),
+    DROP_SHADOW_WIDTH(), DROP_SHADOW_HEIGHT()};
+}
+
+InfoTip::Orientation InfoTip::get_orientation() const {
+  auto parent_position = parentWidget()->mapToGlobal(
+    parentWidget()->rect().bottomLeft());
+  auto screen_geometry =
+    get_current_screen(parent_position)->availableGeometry();
+  if((parent_position.y() + height()) >
```

Comment:
Same thing here, too many parenthesis, stick to C++'s order of operations, you can reference it here:

https://en.cppreference.com/w/cpp/language/operator_precedence